src/Processors/Formats/Impl/Parquet/Write.cpp

```diff
@@ -908,97 +995,138 @@ void writeColumnChunkBody(ColumnChunkWriteState & s, const WriteOptions & option
 {
+    chassert(file.offset == 0);
     /// Write the magic bytes. We're a wizard now.
     out.write("PAR1", 4);
+    file.offset = 4;
 }
 
-parq::ColumnChunk finalizeColumnChunkAndWriteFooter(
-    size_t offset_in_file, ColumnChunkWriteState s, const WriteOptions &, WriteBuffer & out)
+void finalizeColumnChunkAndWriteFooter(ColumnChunkWriteState s, FileWriteState & file, WriteBuffer & out)
 {
+    for (auto & location : s.indexes.offset_index.page_locations)
+        location.offset += file.offset;
+
     if (s.column_chunk.meta_data.data_page_offset != -1)
-        s.column_chunk.meta_data.data_page_offset += offset_in_file;
+        s.column_chunk.meta_data.data_page_offset += file.offset;
     if (s.column_chunk.meta_data.__isset.dictionary_page_offset)
-        s.column_chunk.meta_data.dictionary_page_offset += offset_in_file;
-    s.column_chunk.file_offset = offset_in_file + s.column_chunk.meta_data.total_compressed_size;
+        s.column_chunk.meta_data.dictionary_page_offset += file.offset;
+
+    file.offset += s.column_chunk.meta_data.total_compressed_size;
+    s.column_chunk.file_offset = file.offset;
 
-    serializeThriftStruct(s.column_chunk, out);
+    size_t footer_size = serializeThriftStruct(s.column_chunk, out);
+    file.offset += footer_size;
 
-    return s.column_chunk;
+    file.unflushed_bloom_filter_bytes += s.indexes.bloom_filter_data.size() * sizeof(s.indexes.bloom_filter_data[0]);
+    file.current_row_group.column_indexes.push_back(std::move(s.indexes));
+    file.current_row_group.row_group.columns.push_back(std::move(s.column_chunk));
 }
 
-parq::RowGroup makeRowGroup(std::vector<parq::ColumnChunk> column_chunks, size_t num_rows)
+static void flushBloomFilters(FileWriteState & file, WriteBuffer & out)
 {
-    parq::RowGroup r;
+    while (file.row_groups_with_flushed_bloom_filter < file.completed_row_groups.size())
+    {
+        auto & rg = file.completed_row_groups[file.row_groups_with_flushed_bloom_filter];
+        chassert(rg.column_indexes.size() == rg.row_group.columns.size());
+        for (size_t col_idx = 0; col_idx < rg.column_indexes.size(); ++col_idx)
+        {
+            auto & indexes = rg.column_indexes[col_idx];
+            if (indexes.bloom_filter_data.empty())
+                continue;
+            auto & column = rg.row_group.columns[col_idx];
+            column.meta_data.__set_bloom_filter_offset(file.offset);
+
+            size_t header_size = serializeThriftStruct(indexes.bloom_filter_header, out);
+            size_t data_size = indexes.bloom_filter_data.size() * sizeof(indexes.bloom_filter_data[0]);
+            out.write(indexes.bloom_filter_data.raw_data(), data_size);
+
+            column.meta_data.__set_bloom_filter_length(Int32(header_size + data_size));
+            file.offset += column.meta_data.bloom_filter_length;
+
+            indexes.bloom_filter_data = {}; // free the memory
+        }
+        file.row_groups_with_flushed_bloom_filter += 1;
```

Comment:
what is the point of `row_groups_with_flushed_bloom_filter`? 
I don't see it used anywhere, it's only a local counter here

Reply:
`flushBloomFilters` can be called multiple times (e.g. after each row group), and `file.completed_row_groups.size()` increases between the calls.